### PR TITLE
Feature: add support for a Tiled Gallery layout (Square Tiles) as a Gutenberg block

### PR DESCRIPTION
**Pull request by haszari**
Related: #8527 (see limitations below)

#### Changes proposed in this Pull Request:

Currently, there's no way to use Jetpack Tiled Galleries in Gutenberg. This pull request implements a new Gutenberg block which provides one of the Tiled Gallery layouts as a Gutenberg block. This block can be easily transformed to and from the core Gutenberg gallery block.

This is a proof-of-concept which demonstrates one way to provide Tiled Gallery features in Jetpack. There's lots more work to do!

A fundamental question is how Jetpack-powered gallery features should be made available to users – as an extension to `core/gallery`, or as a separate block (or some other approach). This PR is a demonstration of how this can work well as a separate block.

#### Testing instructions:

I have a sandbox running a production build of this branch - ping me if you want an account to test.
https://haszari.wpsandbox.me

Otherwise, to test this on your WordPress site....

1. Ensure both Gutenberg and Jetpack (this branch!) are installed.
1. Activate & connect Jetpack and enable the "Serve images from our servers" (Photon) option. This option must be enabled to enable Tiled Galleries.
1. Edit a post using Gutenberg.

You can add Jetpack Gallery blocks using the Gutenberg switcher. The interface is somewhat similar to the core block experience, with live preview of the Square Tiles layout, and using the (legacy) modal for adding or editing the images and captions. There are settings for number of columns and what the images should link to (attachment, file, or no link).

screen shot 2018-04-28 at 3 25 31 pm

Using the block switcher you can transform any core gallery to a tiled gallery, and vice versa.

screen shot 2018-04-28 at 3 24 45 pm

#### Limitations/Notes

- Only one tiling algorithm supported at present. If we think this approach is sound, can look at porting other tiling options as follow-up PRs.
- Currently uses a hybrid of new react-generated markup alongside the existing jQuery-based resize handling script (and existing CSS). Longer term it would make sense for this behaviour to be integrated into the react components.
- Is heavily based on the current `core/gallery` implementation, especially for the edit component. This delegates image selection, ordering etc to the legacy edit gallery modal. Ideally the uploading and reordering of images (and other core editing) would be implemented within Gutenberg (WYSIWYG).
- Currently captions are editable in the gallery edit modal (in contrast to `core/gallery`, which has implemented direct manipulation).
- The "image size" option is not correctly handled (in keeping with `core/gallery`).
- Images render squished when previewing in Gutenberg. In development I had thought this was intended behaviour for Square Tiles (had seen it when testing the base plugin). As it appears to render with the correct pixel aspect ratio on the front-end, this should probably be fixed.
- A hard-coded value is used for `content_width`. I suspect this means that tiled galleries will not slot in beautifully in some themes. This value is not currently available to the js/front end (that I'm aware of). Is this heavily used? If so, we can look at ways to make it available to the client side.
- The jsx code for this plugin is built as a new entry point in webpack. This works fine for now, but long term it would be much better to bundle all of Jetpack's Gutenberg jsx as a single file.

### CODE REVIEW

**Comment by jeherve** [comment]
Related: #8527

**Comment by haszari** [comment]
Looks like the CI build started failing when I changed some comments? Is there something else going on here? (I'm pretty certain it passed originally.)

Can that be easily rerun?

**Comment by sirreal** [comment]
> Can that be easily rerun?

I reran the 5.3 tests which had failed. All green now 👌

**Comment by oskosk** [comment]
Given that this PR comes from a fork of the Jetpack repo, I felt the need to drop the instructions on how to achieve a local checkout of this branch for those working with a direct clone of Automattic/Jetpack.

To check this PR:

```
git fetch origin pull/9434/head:pr/9434
git checkout pr/9434
yarn build
```

Make sure Gutenberg is installed and Jetpack is connected.

File: modules/tiled-gallery/tiled-gallery.php [hunk 1]
```diff
@@ -228,6 +228,26 @@ static function get_talaveras() {
 		return self::$talaveras;
 	}
 
+	/**
+ 	 * Enqueue js for our jetpack/gallery block
+ 	 */
+ 	public static function enqueue_block_editor_assets() {
+  		wp_register_script(
+  			'jetpack-tiled-gallery-block',
+  			plugins_url( '../../_inc/build/modules-tiled-gallery-block.js', __FILE__ ), // this is built as a new webpack entry point
```
Comment (tyxla) [inline modules/tiled-gallery/tiled-gallery.php]: Instead of `__FILE__` here we could use `JETPACK__PLUGIN_FILE` to avoid the `../../` relative paths.

Comment (haszari) [reply modules/tiled-gallery/tiled-gallery.php]: That's more robust, thanks. Now fixed :)

File: modules/tiled-gallery/block-edit.jsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+			)
+		);
+
+		if ( images.length === 0 ) {
+			return [
+				controls,
+				<ImagePlaceholder key="gallery-placeholder"
+					className={ className }
+					icon="format-gallery"
+					label={ __( 'Gallery' ) }
```
[the remark this answers is not in the document]
Comment (haszari) [reply modules/tiled-gallery/block-edit.jsx]: Good point! I wish I'd spotted that earlier!

**Comment by oskosk** [comment]
@haszari could you create a version of this branch but on this same repo instead of being a branch on a fork ?

**Comment by haszari** [comment]
@oskosk Will do, thanks for following up. It will be interesting to see how stale this code is – hasn't been rebased in a while. Will let you know when it's up.

**Comment by ockham** [comment]
I"m considering playing with this as a testbed for the Gutenberg/Calypso integration project, so I might as well do it.

**Comment by oskosk** [comment]
Closing in favour of #9903

**Pull request closed without merging**